Source/Dafny/Cloner.cs

```diff
@@ -684,7 +684,7 @@ class Cloner
     public ExtendedPattern CloneExtendedPattern(ExtendedPattern pat) {
       if(pat is LitPattern) {
         var p = (LitPattern)pat;
-        return new LitPattern(p.Tok, (LiteralExpr)CloneExpr(p.Lit));
+        return new LitPattern(p.Tok, CloneExpr(p.OrigLit));
```

Comment:
If CloneExpr is really producing a clone, shouldn't it copy both Lit and OrigLit?  (I wish every AST node has its own ShallowCopy and DeepCopy methods-- would make maintenance/extension more robust)

Reply:
The standard Clone routines output AST nodes with the same information as after parsing. This lets resolution recompute the remaining fields.